In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

os.chdir('/Users/karla/SF-dat/SF-DAT-20/Data/') 
data = pd.read_csv('titanic.csv')
data.head()
### ALL ANSWERS AT THE BOTTOM 

survived  pclass                                               name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      sex  age  sibsp  parch            ticket     fare cabin embarked  
0    male   22      1      0         A/5 21171   7.2500   NaN        S  
1  female   38      1      0          PC 17599  71.2833   C85        C  
2  female   26      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female   35      1      0            113803  53.1000  C123        S  
4    male   35      0      0            373450   8.0500   NaN        S

1. How many passengers are in our passenger list? From here forward, we’ll assume our dataset represents the full passenger list for the Titanic.
2. What is the overall survival rate?  ALL ANSWERS AT THE BOTTOM

In [3]:
data.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [18]:
data['IsFemale'] = data.sex.map({'male':0, 'female':1})
F=data.IsFemale.sum()
print F, data.IsFemale.count()-F
data['HasParch'] = np.where(data.parch > 0, 1, 0)
data.groupby(['sex']).describe()



314 577


HasParch  IsFemale         age        fare       parch  \
sex                                                                      
female count  314.000000       314  261.000000  314.000000  314.000000   
       mean     0.382166         1   27.915709   44.479818    0.649682   
       std      0.486692         0   14.110146   57.997698    1.022846   
       min      0.000000         1    0.750000    6.750000    0.000000   
       25%      0.000000         1   18.000000   12.071875    0.000000   
       50%      0.000000         1   27.000000   23.000000    0.000000   
       75%      1.000000         1   37.000000   55.000000    1.000000   
       max      1.000000         1   63.000000  512.329200    6.000000   
male   count  577.000000       577  453.000000  577.000000  577.000000   
       mean     0.161179         0   30.726645   25.523893    0.235702   
       std      0.368015         0   14.678201   43.138263    0.612294   
       min      0.000000         0    0.420000    0.000000    0.000000   
       25%      0.000000         0   21.000000    7.895800    0.000000   
       50%      0.000000         0   29.000000   10.500000    0.000000   
       75%      0.000000         0   39.000000   26.550000    0.000000   
       max      1.000000         0   80.000000  512.329200    5.000000   

                  pclass       sibsp    survived  
sex                                               
female count  314.000000  314.000000  314.000000  
       mean     2.159236    0.694268    0.742038  
       std      0.857290    1.156520    0.438211  
       min      1.000000    0.000000    0.000000  
       25%      1.000000    0.000000    0.000000  
       50%      2.000000    0.000000    1.000000  
       75%      3.000000    1.000000    1.000000  
       max      3.000000    8.000000    1.000000  
male   count  577.000000  577.000000  577.000000  
       mean     2.389948    0.429809    0.188908  
       std      0.813580    1.061811    0.391775  
       min      1.000000    0.000000    0.000000  
       25%      2.000000    0.000000    0.000000  
       50%      3.000000    0.000000    0.000000  
       75%      3.000000    0.000000    0.000000  
       max      3.000000    8.000000    1.000000

In [19]:
data.groupby(['survived']).describe()

HasParch    IsFemale         age        fare       parch  \
survived                                                                     
0        count  549.000000  549.000000  424.000000  549.000000  549.000000   
         mean     0.189435    0.147541   30.626179   22.117887    0.329690   
         std      0.392211    0.354968   14.172110   31.388207    0.823166   
         min      0.000000    0.000000    1.000000    0.000000    0.000000   
         25%      0.000000    0.000000   21.000000    7.854200    0.000000   
         50%      0.000000    0.000000   28.000000   10.500000    0.000000   
         75%      0.000000    0.000000   39.000000   26.000000    0.000000   
         max      1.000000    1.000000   74.000000  263.000000    6.000000   
1        count  342.000000  342.000000  290.000000  342.000000  342.000000   
         mean     0.318713    0.681287   28.343690   48.395408    0.464912   
         std      0.466660    0.466660   14.950952   66.596998    0.771712   
         min      0.000000    0.000000    0.420000    0.000000    0.000000   
         25%      0.000000    0.000000   19.000000   12.475000    0.000000   
         50%      0.000000    1.000000   28.000000   26.000000    0.000000   
         75%      1.000000    1.000000   36.000000   57.000000    1.000000   
         max      1.000000    1.000000   80.000000  512.329200    5.000000   

                    pclass       sibsp  
survived                                
0        count  549.000000  549.000000  
         mean     2.531876    0.553734  
         std      0.735805    1.288399  
         min      1.000000    0.000000  
         25%      2.000000    0.000000  
         50%      3.000000    0.000000  
         75%      3.000000    1.000000  
         max      3.000000    8.000000  
1        count  342.000000  342.000000  
         mean     1.950292    0.473684  
         std      0.863321    0.708688  
         min      1.000000    0.000000  
         25%      1.000000    0.000000  
         50%      2.000000    0.000000  
         75%      3.000000    1.000000  
         max      3.000000    4.000000

In [20]:
data.groupby(['pclass']).describe()

HasParch    IsFemale         age        fare       parch  \
pclass                                                                     
1      count  216.000000  216.000000  186.000000  216.000000  216.000000   
       mean     0.245370    0.435185   38.233441   84.154687    0.356481   
       std      0.431306    0.496933   14.802856   78.380373    0.693997   
       min      0.000000    0.000000    0.920000    0.000000    0.000000   
       25%      0.000000    0.000000   27.000000   30.923950    0.000000   
       50%      0.000000    0.000000   37.000000   60.287500    0.000000   
       75%      0.000000    1.000000   49.000000   93.500000    0.000000   
       max      1.000000    1.000000   80.000000  512.329200    4.000000   
2      count  184.000000  184.000000  173.000000  184.000000  184.000000   
       mean     0.271739    0.413043   29.877630   20.662183    0.380435   
       std      0.446070    0.493724   14.001077   13.417399    0.690963   
       min      0.000000    0.000000    0.670000    0.000000    0.000000   
       25%      0.000000    0.000000   23.000000   13.000000    0.000000   
       50%      0.000000    0.000000   29.000000   14.250000    0.000000   
       75%      1.000000    1.000000   36.000000   26.000000    1.000000   
       max      1.000000    1.000000   70.000000   73.500000    3.000000   
3      count  491.000000  491.000000  355.000000  491.000000  491.000000   
       mean     0.224033    0.293279   25.140620   13.675550    0.393075   
       std      0.417369    0.455730   12.495398   11.778142    0.888861   
       min      0.000000    0.000000    0.420000    0.000000    0.000000   
       25%      0.000000    0.000000   18.000000    7.750000    0.000000   
       50%      0.000000    0.000000   24.000000    8.050000    0.000000   
       75%      0.000000    1.000000   32.000000   15.500000    0.000000   
       max      1.000000    1.000000   74.000000   69.550000    6.000000   

                   sibsp    survived  
pclass                                
1      count  216.000000  216.000000  
       mean     0.416667    0.629630  
       std      0.611898    0.484026  
       min      0.000000    0.000000  
       25%      0.000000    0.000000  
       50%      0.000000    1.000000  
       75%      1.000000    1.000000  
       max      3.000000    1.000000  
2      count  184.000000  184.000000  
       mean     0.402174    0.472826  
       std      0.601633    0.500623  
       min      0.000000    0.000000  
       25%      0.000000    0.000000  
       50%      0.000000    0.000000  
       75%      1.000000    1.000000  
       max      3.000000    1.000000  
3      count  491.000000  491.000000  
       mean     0.615071    0.242363  
       std      1.374883    0.428949  
       min      0.000000    0.000000  
       25%      0.000000    0.000000  
       50%      0.000000    0.000000  
       75%      1.000000    0.000000  
       max      8.000000    1.000000

In [21]:
data.groupby(['pclass','sex']).describe()

HasParch  IsFemale         age        fare       parch  \
pclass sex                                                                      
1      female count   94.000000        94   85.000000   94.000000   94.000000   
              mean     0.319149         1   34.611765  106.125798    0.457447   
              std      0.468646         0   13.612052   74.259988    0.728305   
              min      0.000000         1    2.000000   25.929200    0.000000   
              25%      0.000000         1   23.000000   57.244800    0.000000   
              50%      0.000000         1   35.000000   82.664550    0.000000   
              75%      1.000000         1   44.000000  134.500000    1.000000   
              max      1.000000         1   63.000000  512.329200    2.000000   
       male   count  122.000000       122  101.000000  122.000000  122.000000   
              mean     0.188525         0   41.281386   67.226127    0.278689   
              std      0.392743         0   15.139570   77.548021    0.658853   
              min      0.000000         0    0.920000    0.000000    0.000000   
              25%      0.000000         0   30.000000   27.728100    0.000000   
              50%      0.000000         0   40.000000   41.262500    0.000000   
              75%      0.000000         0   51.000000   78.459375    0.000000   
              max      1.000000         0   80.000000  512.329200    4.000000   
2      female count   76.000000        76   74.000000   76.000000   76.000000   
              mean     0.407895         1   28.722973   21.970121    0.605263   
              std      0.494709         0   12.872702   10.891796    0.833930   
              min      0.000000         1    2.000000   10.500000    0.000000   
              25%      0.000000         1   22.250000   13.000000    0.000000   
              50%      0.000000         1   28.000000   22.000000    0.000000   
              75%      1.000000         1   36.000000   26.062500    1.000000   
              max      1.000000         1   57.000000   65.000000    3.000000   
       male   count  108.000000       108   99.000000  108.000000  108.000000   
              mean     0.175926         0   30.740707   19.741782    0.222222   
              std      0.382532         0   14.793894   14.922235    0.517603   
              min      0.000000         0    0.670000    0.000000    0.000000   
              25%      0.000000         0   23.000000   12.331250    0.000000   
              50%      0.000000         0   30.000000   13.000000    0.000000   
              75%      0.000000         0   36.750000   26.000000    0.000000   
              max      1.000000         0   70.000000   73.500000    2.000000   
3      female count  144.000000       144  102.000000  144.000000  144.000000   
              mean     0.409722         1   21.750000   16.118810    0.798611   
              std      0.493499         0   12.729964   11.690314    1.237976   
              min      0.000000         1    0.750000    6.750000    0.000000   
              25%      0.000000         1   14.125000    7.854200    0.000000   
              50%      0.000000         1   21.500000   12.475000    0.000000   
              75%      1.000000         1   29.750000   20.221875    1.000000   
              max      1.000000         1   63.000000   69.550000    6.000000   
       male   count  347.000000       347  253.000000  347.000000  347.000000   
              mean     0.146974         0   26.507589   12.661633    0.224784   
              std      0.354591         0   12.159514   11.681696    0.623404   
              min      0.000000         0    0.420000    0.000000    0.000000   
              25%      0.000000         0   20.000000    7.750000    0.000000   
              50%      0.000000         0   25.000000    7.925000    0.000000   
              75%      0.000000         0   33.000000   10.008300    0.000000   
              max      1.000000         0   74.000000 

In [26]:
data[data['pclass']==3].groupby(['HasParch','sex']).describe()

IsFemale         age        fare       parch  pclass  \
HasParch sex                                                                  
0        female count        85   54.000000   85.000000   85.000000      85   
                mean          1   24.018519   10.207652    0.000000       3   
                std           0    9.697774    3.918115    0.000000       0   
                min           1    5.000000    6.750000    0.000000       3   
                25%           1   18.000000    7.750000    0.000000       3   
                50%           1   22.000000    7.925000    0.000000       3   
                75%           1   27.750000   12.475000    0.000000       3   
                max           1   63.000000   24.150000    0.000000       3   
         male   count       296  210.000000  296.000000  296.000000     296   
                mean          0   28.816667    9.970506    0.000000       3   
                std           0   10.445963    8.047095    0.000000       0   
                min           0   11.000000    0.000000    0.000000       3   
                25%           0   21.000000    7.732275    0.000000       3   
                50%           0   26.000000    7.895800    0.000000       3   
                75%           0   33.000000    8.315625    0.000000       3   
                max           0   74.000000   56.495800    0.000000       3   
1        female count        59   48.000000   59.000000   59.000000      59   
                mean          1   19.197917   24.634885    1.949153       3   
                std           0   15.153291   13.774286    1.223671       0   
                min           1    0.750000    7.750000    1.000000       3   
                25%           1    4.750000   15.245800    1.000000       3   
                50%           1   17.500000   21.075000    2.000000       3   
                75%           1   30.250000   31.275000    2.000000       3   
                max           1   48.000000   69.550000    6.000000       3   
         male   count        51   43.000000   51.000000   51.000000      51   
                mean          0   15.230698   28.280722    1.529412       3   
                std           0   13.684835   16.448391    0.808775       0   
                min           0    0.420000    7.229200    1.000000       3   
                25%           0    4.000000   15.820850    1.000000       3   
                50%           0   10.000000   25.466700    1.000000       3   
                75%           0   19.000000   37.031250    2.000000       3   
                max           0   44.000000   69.550000    5.000000       3   

                            sibsp    survived  
HasParch sex                                   
0        female count   85.000000   85.000000  
                mean     0.341176    0.588235  
                std      0.588641    0.495074  
                min      0.000000    0.000000  
                25%      0.000000    0.000000  
                50%      0.000000    1.000000  
                75%      1.000000    1.000000  
                max      3.000000    1.000000  
         male   count  296.000000  296.000000  
                mean     0.135135    0.121622  
                std      0.414134    0.327402  
                min      0.000000    0.000000  
                25%      0.000000    0.000000  
                50%      0.000000    0.000000  
                75%      0.000000    0.000000  
                max      2.000000    1.000000  
1        female count   59.000000   59.000000  
                mean     1.694915    0.372881  
                std      2.044794    0.487722  
                min      0.000000    0.000000  
                25%      0.000000    0.000000  
                50%      1.000000    0.000000  
                75%      3.000000    1.000000  
                max      8.000000    1.000000  
         male   count   51.000000   51.000000  
                mean     2.607843    0.2

1. How many passengers are in our passenger list? From here forward, we’ll assume our dataset represents the full passenger list for the Titanic.

891 passengers

2. What is the overall survival rate?
38.38% of passengers survived

3. How many male passengers were onboard?
577

4. How many female passengers were onboard? 
314


5. What is the overall survival rate of male passengers?
18.89%


6. What is the overall survival rate of female passengers?
74.2%


7. What is the average age of all passengers onboard? * How did you calculate this average age?
29.69 years old calculated with the describe function.

* Note that some of the passengers do not have an age value. How did you deal with this? What are some other ways of dealing with this?
The describe() function handles a missing value by excluding the observation from the calculation of the mean.
Other ways to deal with the missing values would be to create a dataset and drop the observations where there is no age value, or run a python calculation which takes the sum of only non missing values and divides by the number of items in the sum.  But it seems sensible to me to use a builtin and tested method.
    
  
8. What is the average age of passengers who survived?
28.34

9. What is the average age of passengers who did not survive?
30.26

10. At this (early) point in our analysis, what might you infer about any patterns you are seeing?  
Gender was the biggest influence on survival rate, bigger than age.  More females survived.

11. How many passengers are in each of the three classes of service (e.g. First, Second, Third?)
216, 184, 491

12. What is the survival rate for passengers in each of the three classes of service?
1,2,3 = 63%, 47%, 24%

13. What else might you conclude?
It was better to be in first class. (97% of women survived, 37% of men)
It was better to be female over all. (97% in first, 92% in second, 50% in third) vs (37%, 16%, 14%).
Avoid travelling third class.

A higher percentage of third class was male than in the other two classes. (56%, 59%, 70%).
The survival rate for men in second and third class is very similar(16%, 14%), which is not the case for females (92%,50%).


14. Last, if we were to build a predictive model, which features in the data do you think we should include in the model and which can we leave out? Why?  

I would use gender and class in the predictive model. 

The model might get some fine tuning by knowing whether a person had family on board or not. For instance, what is the survival rate for women who are mothers vs those who are not? This affect might be seen especially in third class, where so many more of the women died than in the other two classes.  And for family men?  The sibsp and parch columns would be useful for this.  

Age is evenly spread between survival groups and within gender groups so not useful for gross prediction purposes, but it might help analyzing family related survival factors. Sadly, looking at parent/child yes or no, looking at third class only, the survival rate for females who have a parch=0 (not a child or parent) was higher than for those who were either a child or a parent.

The cabin numbers are too sparsely recorded to accurately identify traveling companions. I don't think embarkation point is relevant to predicting survival rate.

Fare price reflects class. 

